napari/_vispy/canvas.py

```diff
@@ -578,10 +600,22 @@ def add_layer_visual_mapping(
 
         vispy_layer.node.parent = self.view.scene
         self.layer_to_visual[napari_layer] = vispy_layer
+        self._layer_overlay_to_visual[napari_layer] = {}
 
         napari_layer.events.visible.connect(self._reorder_layers)
+        overlay_callback = partial(self._update_layer_overlays, napari_layer)
+        napari_layer._overlays.events.added.connect(overlay_callback)
+        napari_layer._overlays.events.removed.connect(overlay_callback)
+        napari_layer._overlays.events.changed.connect(overlay_callback)
+        self._overlay_callbacks[napari_layer] = overlay_callback
```

Comment:
This was the best solution I could find in order to be able to disconnect this properly later. The issue is that the overlay events have no knowledge about the *layer* they belong to, and by using partials we need to keep the callback around to be able to disconnect it.

@Czaki I know you dealt with this partials+connect before, maybe you have a better idea?